cluster/job_once_scheduler.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cluster
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// syncedCallback uses the mutex to make things predictable for the client: the callback will be
+// called once at a time (the client does not need to worry about concurrency within the callback)
+type syncedCallback struct {
+	mu       sync.Mutex
```

Comment:
It would be sufficient for the same job, but not for multiple jobs.

Reply:
Are there any constraints on the API consumer? i.e. Does the callback need to be synchronous?

Reply:
That's a constraint I wanted to impose to keep things predictable for the client. I don't want the plugins using this to have to worry about concurrency inside their callback handler.

Reply:
But for multiple jobs there would be multiple mutexes so no synchronization? What is the situation we are avoiding here?

Reply:
All the jobs call the same callback, so two different jobs can be locked with their own mutexes and still call the callback concurrently. It might be rare, but definitely possible.

Reply:
Ohhhhhhh I see. 👍